File: ai_service/app/models/ai_token_usage.py
"""
SQLAlchemy model for tracking AI API token usage.
This model connects to the admin-core-service database.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import uuid4

from sqlalchemy import (
    Column,
    String,
    Integer,
    DateTime,
    Index,
    text,
    Enum as SQLEnum,
    Numeric,
    TypeDecorator,
)
from sqlalchemy.dialects.postgresql import UUID
import enum

from .ai_gen_video import Base


class EnumValueType(TypeDecorator):
    """Type decorator to ensure enum values (not names) are stored in database."""
    impl = String
    cache_ok = True
    
    def __init__(self, enum_class, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.enum_class = enum_class
    
    def process_bind_param(self, value, dialect):
        """Convert enum to its value (string) when binding to database."""
        if value is None:
            return None
        
        # Handle enum instances - get the value
        if isinstance(value, enum.Enum):
            return value.value
        
        # Handle strings - check if it's an enum name or value
        if isinstance(value, str):
            # First check if it's already a valid enum value (lowercase)
            valid_values = [e.value for e in self.enum_class]
            if value.lower() in [v.lower() for v in valid_values]:
                # Return the correct lowercase value
                for v in valid_values:
                    if v.lower() == value.lower():
                        return v
            
            # If it's an enum name (uppercase), convert to value
            try:
                # Try to find enum by name (case-insensitive)
                for enum_member in self.enum_class:
                    if enum_member.name.upper() == value.upper():
                        return enum_member.value
                # Try direct lookup
                enum_member = self.enum_class[value]
                return enum_member.value
            except (KeyError, AttributeError):
                # If not found, return lowercase version if it looks like an enum name
                if value.isupper() and len(value) > 1:
                    # Might be an enum name, try to convert
                    try:
                        return self.enum_class[value].value
                    except KeyError:
                        pass
                return value.lower() if value else value
        
        # For other types, convert to string and lowercase
        return str(value).lower() if value else None
    
    def process_result_value(self, value, dialect):
        """Convert database value back to enum when reading."""
        if value is None:
            return None
        try:
            return self.enum_class(value)
        except ValueError:
            return value


class ApiProvider(str, enum.Enum):
    """Enum for API provider types."""
    OPENAI = "openai"
    GEMINI = "gemini"


class RequestType(str, enum.Enum):
    """Enum for request types."""
    OUTLINE = "outline"
    IMAGE = "image"
    CONTENT = "content"
    VIDEO = "video"


class AiTokenUsage(Base):
    """
    Tracks token usage for AI API calls (OpenAI/OpenRouter and Gemini).
    
    Records token consumption per request to enable:
    - Usage tracking and billing
    - Cost analysis
    - Rate limiting
    - Usage reporting
    """
    __tablename__ = "ai_token_usage"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4, server_default=text("gen_random_uuid()"))
    
    # Entity identification (optional, for tracking usage per institute/user)
    institute_id = Column(UUID(as_uuid=True), nullable=True, index=True)
    user_id = Column(UUID(as_uuid=True), nullable=True, index=True)
    
    # API provider and model information
    # Use EnumValueType to ensure enum values (lowercase strings) are stored, not enum names
    api_provider = Column(EnumValueType(ApiProvider, 50), nullable=False, index=True)
    model = Column(String(255), nullable=True, index=True)
    
    # Token usage metrics
    # Note: prompt_tokens = input tokens, completion_tokens = output tokens
    prompt_tokens = Column(Integer, nullable=False, default=0, comment="Input tokens (prompt tokens)")
    completion_tokens = Column(Integer, nullable=False, default=0, comment="Output tokens (completion tokens)")
    total_tokens = Column(Integer, nullable=False, default=0, index=True)
    
    # Pricing information (per token pricing for input and output)
    # input_token_price applies to prompt_tokens, output_token_price applies to completion_tokens
    input_token_price = Column(Numeric(20, 10), nullable=True, comment="Price per input token (applies to prompt_tokens) for this model")
    output_token_price = Column(Numeric(20, 10), nullable=True, comment="Price per output token (applies to completion_tokens) for this model")
    total_price = Column(Numeric(20, 10), nullable=True, comment="Total price: (input_token_price * prompt_tokens) + (output_token_price * completion_tokens)")
    
    # Request context
    # Use EnumValueType to ensure enum values (lowercase strings) are stored, not enum names
    request_type = Column(EnumValueType(RequestType, 50), nullable=False, index=True)
    request_id = Column(String(255), nullable=True, index=True, comment="Optional request identifier for correlation")
    
    # Additional metadata (Python attribute is 'request_metadata' to avoid SQLAlchemy reserved name conflict,
    # but maps to database column 'metadata' for backward compatibility)
    request_metadata = Column("metadata", String(500), nullable=True, comment="Optional JSON string for additional context")
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow, index=True)
    
    def __repr__(self) -> str:
        entity = f"institute_id={self.institute_id}" if self.institute_id else f"user_id={self.user_id}"
        return f"<AiTokenUsage({entity}, provider={self.api_provider}, tokens={self.total_tokens})>"
    
    def to_dict(self) -> dict:
        """Convert model to dictionary for API responses."""
        return {
            "id": str(self.id),
            "institute_id": str(self.institute_id) if self.institute_id else None,
            "user_id": str(self.user_id) if self.user_id else None,
            "api_provider": self.api_provider.value if self.api_provider else None,
            "model": self.model,
            "prompt_tokens": self.prompt_tokens,
            "completion_tokens": self.completion_tokens,
            "total_tokens": self.total_tokens,
            "input_token_price": float(self.input_token_price) if self.input_token_price is not None else None,
            "output_token_price": float(self.output_token_price) if self.output_token_price is not None else None,
            "total_price": float(self.total_price) if self.total_price is not None else None,
            "request_type": self.request_type.value if self.request_type else None,
            "request_id": self.request_id,
            "request_metadata": self.request_metadata,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }


# Create indexes for efficient queries
Index("idx_ai_token_usage_institute_created", AiTokenUsage.institute_id, AiTokenUsage.created_at)
Index("idx_ai_token_usage_user_created", AiTokenUsage.user_id, AiTokenUsage.created_at)
Index("idx_ai_token_usage_provider_created", AiTokenUsage.api_provider, AiTokenUsage.created_at)
Index("idx_ai_token_usage_type_created", AiTokenUsage.request_type, AiTokenUsage.created_at)

File: admin_core_service/src/main/java/vacademy/io/admin_core_service/features/ai_usage/entity/AiTokenUsage.java
package vacademy.io.admin_core_service.features.ai_usage.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.math.BigDecimal;
import java.time.ZonedDateTime;
import java.util.UUID;

/**
 * Entity for tracking AI token usage across all services.
 * This table logs usage for OpenRouter, Gemini, and Google TTS APIs.
 * 
 * Note: The table is created via Flyway migrations, not JPA auto-generate.
 */
@Entity
@Table(name = "ai_token_usage", indexes = {
        @Index(name = "idx_ai_token_usage_institute_id", columnList = "institute_id"),
        @Index(name = "idx_ai_token_usage_user_id", columnList = "user_id"),
        @Index(name = "idx_ai_token_usage_api_provider", columnList = "api_provider"),
        @Index(name = "idx_ai_token_usage_request_type", columnList = "request_type"),
        @Index(name = "idx_ai_token_usage_created_at", columnList = "created_at")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class AiTokenUsage {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "institute_id")
    private UUID instituteId;

    @Column(name = "user_id")
    private UUID userId;

    /**
     * API provider: openai, gemini
     * OpenRouter calls are logged as 'openai' since they use OpenAI-compatible API
     */
    @Column(name = "api_provider", nullable = false, length = 50)
    private String apiProvider;

    /**
     * Model name, e.g., "google/gemini-2.5-flash", "openai/gpt-4o-mini"
     */
    @Column(name = "model", length = 255)
    private String model;

    /**
     * Number of input/prompt tokens
     */
    @Column(name = "prompt_tokens", nullable = false)
    @Builder.Default
    private Integer promptTokens = 0;

    /**
     * Number of output/completion tokens
     */
    @Column(name = "completion_tokens", nullable = false)
    @Builder.Default
    private Integer completionTokens = 0;

    /**
     * Total tokens (prompt_tokens + completion_tokens)
     */
    @Column(name = "total_tokens", nullable = false)
    @Builder.Default
    private Integer totalTokens = 0;

    /**
     * Price per input token for this model
     */
    @Column(name = "input_token_price", precision = 20, scale = 10)
    private BigDecimal inputTokenPrice;

    /**
     * Price per output token for this model
     */
    @Column(name = "output_token_price", precision = 20, scale = 10)
    private BigDecimal outputTokenPrice;

    /**
     * Total calculated price
     */
    @Column(name = "total_price", precision = 20, scale = 10)
    private BigDecimal totalPrice;

    /**
     * Request type: outline, image, content, video, tts, embedding, evaluation,
     * presentation, conversation, lecture, course_content, pdf_questions, agent,
     * analytics, copilot
     */
    @Column(name = "request_type", nullable = false, length = 50)
    private String requestType;

    /**
     * Optional request correlation ID
     */
    @Column(name = "request_id", length = 255)
    private String requestId;

    /**
     * Optional JSON metadata
     */
    @Column(name = "metadata", length = 500)
    private String metadata;

    /**
     * TTS provider for TTS requests: google, edge, elevenlabs
     */
    @Column(name = "tts_provider", length = 50)
    private String ttsProvider;

    /**
     * Character count for TTS requests (TTS is charged by character, not token)
     */
    @Column(name = "character_count")
    private Integer characterCount;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private ZonedDateTime createdAt;

    /**
     * Convenience method to calculate total tokens if not already set.
     */
    @PrePersist
    public void prePersist() {
        if (totalTokens == null || totalTokens == 0) {
            totalTokens = (promptTokens != null ? promptTokens : 0)
                    + (completionTokens != null ? completionTokens : 0);
        }
    }
}
